cmd/query/app/grpcserver/server.go

```diff
@@ -0,0 +1,126 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcserver
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/soheilhy/cmux"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+
+	"github.com/jaegertracing/jaeger/cmd/flags"
+	"github.com/jaegertracing/jaeger/cmd/query/app"
+	"github.com/jaegertracing/jaeger/cmd/query/app/querysvc"
+	"github.com/jaegertracing/jaeger/pkg/healthcheck"
+)
+
+const (
+	grpcListener = "grpc"
+	httpListener = "http"
+)
+
+// Builder provides the configuration used to build a GrpcServer
+type Builder struct {
+	Svc             *flags.Service
+	RecoveryHandler http.Handler
+	QuerySvc        querysvc.QueryService
+	Logger          *zap.Logger
+	Tracer          opentracing.Tracer
+	QueryOptions    *app.QueryOptions
+}
+
+// GrpcServer runs HTTP, Mux and a grpc server
+type GrpcServer struct {
+	svc        *flags.Service
+	grpcServer *grpc.Server
+
+	listeners  map[string]net.Listener
+	httpServer *http.Server
+	muxServer  cmux.CMux
+	logger     *zap.Logger
+	listenPort int
+}
+
+func (g *Builder) createHandler() *app.GRPCHandler {
+	return app.NewGRPCHandler(g.QuerySvc, g.Logger, g.Tracer)
+}
+
+// Build a GrpcServer
+func (g *Builder) Build() *GrpcServer {
+
+	// Prepare cmux conn.
+	conn, err := net.Listen("tcp", fmt.Sprintf(":%d", g.QueryOptions.Port))
+	if err != nil {
+		g.Logger.Fatal("Could not start listener", zap.Error(err))
+	}
+	// Create cmux server.
+	// cmux will reverse-proxy between HTTP and GRPC backends.
+	cmuxServer := cmux.New(conn)
+
+	// Add GRPC and HTTP listeners.
+	listeners := map[string]net.Listener{
+		grpcListener: cmuxServer.Match(
+			cmux.HTTP2HeaderField("content-type", "application/grpc"),
+			cmux.HTTP2HeaderField("content-type", "application/grpc+proto")),
+
+		httpListener: cmuxServer.Match(cmux.Any()),
+	}
+
+	return &GrpcServer{
+		logger:     g.Logger,
+		grpcServer: grpc.NewServer(),
+		muxServer:  cmuxServer,
+		listeners:  listeners,
+		httpServer: &http.Server{
+			Handler: g.RecoveryHandler,
+		},
+	}
+}
+
+// Start http, GRPC and cmux servers concurrently
+func (s *GrpcServer) Start() {
+
+	logger := s.logger // shortcut
+
+	go func() {
+		logger.Info("Starting HTTP server", zap.Int("port", s.listenPort))
+		if err := s.httpServer.Serve(s.listeners[httpListener]); err != nil {
+			logger.Fatal("Could not start HTTP server", zap.Error(err))
+		}
+		s.svc.HC().Set(healthcheck.Unavailable)
```

Comment:
not a goal of this PR, but we should really revisit this, because it's feasible that one of the servers will fail to start and set HC=not-healthy, while the others would start up fine and main will set HC to OK. 

The goroutines should just be signaling to main() that they failed, and main can set the healthcheck.

Maybe add a TODO to fix this separately.

Reply:
In the docs for `logger.Fatal` it says that 
```go
// ....
// The logger then calls os.Exit(1), even if logging at FatalLevel is

```
wouldn't that mean that anything below the log is dead code and won't be executed? If so should we log in the goroutine with `fatal` or just have a mechanism for propagation of the errs to `main()`, where they'll be logged.